test/integration/common/StudioActions.kt

```diff
@@ -72,12 +73,34 @@ object StudioActions {
         clickAllInstancesOfText(composeRule, icon.unicode)
     }
 
-    fun assertNodeExistsWithText(composeRule: ComposeContentTestRule, text: String): SemanticsNodeInteraction {
-        return composeRule.onNodeWithText(text).assertExists()
+    suspend fun waitUntilNodeWithIconIsClickable(composeRule: ComposeContentTestRule, icon: Icon) {
+        waitUntilNodeWithTextIsClickable(composeRule, icon.unicode)
+    }
+
+    suspend fun waitUntilNodeWithTextIsClickable(composeRule: ComposeContentTestRule, text: String) {
+        waitUntilAssertionIsTrue(composeRule) { composeRule.onNodeWithText(text).assertHasClickAction() }
+    }
+
+    suspend fun waitUntilNodeWithTextExists(composeRule: ComposeContentTestRule, text: String) {
+        waitUntilAssertionIsTrue(composeRule) { composeRule.onNodeWithText(text).assertExists() }
+    }
+
+    suspend fun waitUntilNodeWithTextNotExists(composeRule: ComposeContentTestRule, text: String) {
+        waitUntilAssertionIsTrue(composeRule) { composeRule.onNodeWithText(text).assertDoesNotExist() }
     }
 
-    fun assertNodeNotExistsWithText(composeRule: ComposeContentTestRule, text: String) {
-        return composeRule.onNodeWithText(text).assertDoesNotExist()
+    suspend fun waitUntilAssertionIsTrue(composeRule: ComposeContentTestRule, assertion: () -> Any) {
+        composeRule.waitUntil(Delays.WAIT_TIMEOUT) {
+            try {
+                assertion()
+                runBlocking {
+                    composeRule.awaitIdle()
+                }
+                return@waitUntil true
+            } catch (e: Exception) {
```

Comment:
We want this function to be a general one - as such, we expect that our `assertion` parameter could throw any exception under circumstances where it isn't 'ready' yet. Rust's `Fn` trait would be nice here, providing a guarantee that calling the function repeatedly won't mutate state. I don't think there's any equivalent within Kotlin unfortunately.

Reply:
Yeah, you need a borrower-checker for `Fn`. :( 🚀